File: Academy.Backend/src/Shared/Academy.Core/Models/Envelope.cs
using Academy.SharedKernel;

namespace Academy.Core.Models
{
    public record Envelope
    {
        private Envelope(object? result, ErrorList? errors)
        {
            Result = result;
            Errors = errors;
            TimeGenerated = DateTime.UtcNow;
        }

        public object? Result { get; }
        public ErrorList? Errors { get; }
        public DateTime TimeGenerated { get; }

        public static Envelope Ok(object? result = null) => new(result, null);
        public static Envelope Error(ErrorList errors) => new(null, errors);
    }
}

File: Academy.Frontend/src/models/response/Envelope.ts
export interface Envelope<T> {
  result?: T;
  errors?: ErrorList;
  timeGenerated: string;
}

export interface ErrorList {
  errors: Error[];
}

export interface Error {
  code: string;
  message: string;
  type: ErrorType;
  invalidField?: string;
}

export enum ErrorType {
  NotFound = 0,
  Validation = 1,
  Conflict = 2,
  Failure = 3,
  AccessDenied = 4,
  None = 5,
}
